Tools/speech_recognizer/PocketsphinxSpeechRecognitionEngine.m

```diff
@@ -0,0 +1,197 @@
+#import "GSSpeechRecognitionEngine.h"
+#import <Foundation/NSDistributedNotificationCenter.h>
+
+#include <sphinxbase/err.h>
+#include <sphinxbase/ad.h>
+#include <pocketsphinx/pocketsphinx.h>
+
+/**
+ * Implementation of a speech engine using pocketsphinx.  This should be the default
+ * for resource-constrained platforms.
+ */
+
+#define MODELDIR "/share/pocketsphinx/model"
+
+static const arg_t cont_args_def[] = {
+    POCKETSPHINX_OPTIONS,
+    /* Argument file. */
+    {"-argfile",
+     ARG_STRING,
+     NULL,
+     "Argument file giving extra arguments."},
+    {"-adcdev",
+     ARG_STRING,
+     NULL,
+     "Name of audio device to use for input."},
+    {"-infile",
+     ARG_STRING,
+     NULL,
+     "Audio file to transcribe."},
+    {"-inmic",
+     ARG_BOOLEAN,
+     "no",
+     "Transcribe audio from microphone."},
+    {"-time",
+     ARG_BOOLEAN,
+     "no",
+     "Print word times in file transcription."},
+    CMDLN_EMPTY_OPTION
+};
+
+@interface PocketsphinxSpeechRecognitionEngine : GSSpeechRecognitionEngine
+{
+  ps_decoder_t *ps;
+  cmd_ln_t *config;
+  FILE *fh;
+  char const *uttid;
+  int16 buf[512];
+  int rv;
+  int32 score;
+  NSThread *_listeningThread;
+  id<NSSpeechRecognizerDelegate> _delegate;
+}
+@end
+
+@implementation PocketsphinxSpeechRecognitionEngine
+
+- (id)init
+{
+  if ((self = [super init]) != nil)
+    {
+      char *arg[3];
+      arg[0] = "";
+      arg[1] = "-inmic";
+      arg[2] = "yes";
+
+      config = cmd_ln_parse_r(NULL, cont_args_def, 3, arg, TRUE);
+      ps_default_search_args(config);
+      ps = ps_init(config);
+      if (ps == NULL)
+        {
+          cmd_ln_free_r(config);
+          NSLog(@"Could not start server");
+          return nil;
+        }
+      _listeningThread = nil;
+    }
+  return self;
+}
+
+- (void) _recognizedWord: (NSString *)word
+{
+  [[NSDistributedNotificationCenter defaultCenter]
+    postNotificationName: GSSpeechRecognizerDidRecognizeWordNotification
+                  object: word
+                userInfo: nil];
+}
+
+/*
+ * NOTE: This code is derived from continuous.c under pocketsphinx 
+ *       which is MIT licensed
+ * Main utterance processing loop:
+ *     while (YES) {
+ *        start utterance and wait for speech to process
+ *        decoding till end-of-utterance silence will be detected
+ *        print utterance result;
+ *     }
+ */
+- (void) recognize
+{
+  ad_rec_t *ad;
+  int16 adbuf[2048];
+  BOOL utt_started, in_speech;
+  int32 k;
+  char const *hyp;
+
+  NSLog(@"************************* In recognize");
+  if ((ad = ad_open_dev(cmd_ln_str_r(config, "-adcdev"),
+                        (int) cmd_ln_float32_r(config,
+                                               "-samprate"))) == NULL)
+    {
+      NSLog(@"Failed to open audio device");
```

Comment:
Shouldn't all these NSLog cases break out of this method?

Reply:
Corrected.